refactor: client light networking

mickvandijke opened the pull request.

mickvandijke committed:
refactor!: replace `ant_networking` with built-in lightweight network driver

grumbach committed:
feat: make networking its own separate cleanly delimited module

grumbach committed:
chore: fix compile in data location test

mickvandijke committed:
chore(autonomi): restore original pub interface/exports

mickvandijke committed:
chore: cargo fmt

grumbach committed:
fix: adapt retry strategies for existence checks

grumbach committed:
fix: pointer and graph existence checks

grumbach committed:
feat: add logging

grumbach committed:
chore: adjust logging

mickvandijke committed:
Merge pull request #2888 from mickvandijke/refactor-autonomi-use-light-networking-driver3

refactor: autonomi use light networking driver

mickvandijke committed:
chore: use default libp2p settings

mickvandijke committed:
chore: don't set replication factor

mickvandijke committed:
chore: yank get quote retry strategy

mickvandijke committed:
chore: only get quotes from close group

mickvandijke committed:
chore: increase KAD query timeout to 600 seconds

mickvandijke committed:
chore: dial peer before requesting quote

mickvandijke committed:
chore: stop network client from dropping

grumbach committed:
fix: quotes mismanagement issues

grumbach committed:
fix: revive replication factor as it is needed for quorum

grumbach committed:
fix: missing scratchpad case

grumbach committed:
fix: doc comments

mickvandijke committed:
fix(autonomi): assert in `get_closest_n_peers`

More than N results can be returned.

mickvandijke committed:
chore(autonomi): info log client peer id

mickvandijke committed:
chore(autonomi): add identify behaviour

mickvandijke committed:
chore(autonomi): increase resend identify interval

mickvandijke committed:
chore(autonomi): truncate `get_closest_n_peers` response peers to N

mickvandijke committed:
fix(autonomi): get store quote responses from non-local peers

mickvandijke committed:
Merge pull request #8 from mickvandijke/client-light-networking-identify

fix: get store quotes from non-local peers

grumbach committed:
fix: record size limit

mickvandijke committed:
Merge pull request #2891 from mickvandijke/client-light-networking

client light networking

grumbach committed:
fix: pointer split handling

mickvandijke committed:
Merge pull request #2898 from grumbach/client-light-networking-pointer-split

fix: pointer split handling

grumbach committed:
fix: make scratchpad error uniform

grumbach committed:
fix: add a little sleep before get right after put

mickvandijke committed:
chore(autonomi): disable periodic bootstrapping

mickvandijke committed:
fix(autonomi): consider chunk paid when majority has it

grumbach committed:
Merge pull request #2903 from mickvandijke/client-light-networking

fix: disable periodic bootstrap & fix quoting existing records

grumbach committed:
Merge pull request #2900 from grumbach/client-light-networking-fixes

fix: client light networking fixes

grumbach committed:
feat: impl retries for get put and quotes

grumbach committed:
fix: js test and debug assert information

grumbach committed:
fix: make insufficient peers during closest peers search an error

grumbach committed:
feat: improve networking errors

grumbach committed:
chore: add docs for existance check functions

grumbach committed:
feat: get closest peers with retries

grumbach committed:
Merge pull request #2904 from grumbach/client-light-networking-with-retries

feat: impl retries for get put and quotes

grumbach committed:
feat: remove disjoint_query_path

grumbach committed:
feat: retry closest peers

mickvandijke committed:
Merge pull request #2910 from grumbach/client-light-networking-remove-disjoint-query-path

feat: remove disjoint_query_path

mickvandijke committed:
Merge pull request #2911 from grumbach/client-light-networking-closest-peers-retry

feat: retry closest peers

maqi committed:
Revert "test: enable data_with_churn using networkspawner"

This reverts commit bdd34293355d0e6f7b98e0901ba2da268483ed7b.

maqi committed:
Revert "test: use nodespawner for verify_data_location & verify_routing_table"

This reverts commit 90b02d8409dc0720536d6f670362e72c79a899ab.

maqi committed:
fix(client): use libp2p::kad::BucketInserts::OnConnected

grumbach committed:
Merge pull request #2920 from maqi/client-light-networking_resovling_failing_churn_test

grumbach committed:
chore: reduce extra logging in client networking

grumbach committed:
fix: make errors more useful

grumbach committed:
Merge pull request #2923 from grumbach/client-light-networking-reduce-logging

chore: reduce extra logging in client networking

grumbach committed:
Merge pull request #2924 from grumbach/fix-client-error-reporting

fix: make errors more useful

grumbach committed:
Merge branch 'main' into client-light-networking

grumbach committed:
Merge branch 'client-light-networking' of ssh://github.com/maidsafe/autonomi into client-light-networking

grumbach committed:
feat: add easy 1 liner setup for alphanet

grumbach committed:
fix: ignore cache to false

grumbach committed:
Merge pull request #2927 from grumbach/client-light-networking-easy-alpha

feat: add easy 1 liner setup for alphanet

grumbach committed:
Merge branch 'main' into client-light-networking

grumbach committed:
feat: simpler network config for cli

grumbach committed:
feat: make local flag override network id

grumbach committed:
chore: remove useless arg and improve messaging

mickvandijke committed:
Merge pull request #2933 from grumbach/easy_alpha_for_cli

Easy alpha for cli

grumbach committed:
Merge branch 'main' into client-light-networking

grumbach reviewed (approved):
LGTM, since it is quite big, would be good to have another review @maqi or @RolandSherwin?

grumbach committed:
chore: remove unused errors

grumbach committed:
feat: make nodes allow 24 hours instead of 1 hour before quote expires

grumbach committed:
feat: put errors come with a state on put fails to facilitate retries

grumbach committed:
feat: data payments in errors uppon upload network failures

grumbach committed:
feat: trim more unused error variants

grumbach committed:
fix: use universal archive serializer

grumbach committed:
feat: add placeholder for cli integration

mickvandijke committed:
Merge remote-tracking branch 'maidsafe/main' into client-light-networking

# Conflicts:
#	autonomi/Cargo.toml
#	autonomi/src/client/config.rs
#	autonomi/src/client/data_types/chunk.rs
#	autonomi/src/client/data_types/graph.rs
#	autonomi/src/client/data_types/pointer.rs
#	autonomi/src/client/data_types/scratchpad.rs
#	autonomi/src/client/quote.rs

mickvandijke committed:
Merge remote-tracking branch 'maidsafe/main' into client-light-networking-rebased

# Conflicts:
#	ant-cli/src/actions/connect.rs
#	ant-cli/src/commands.rs
#	ant-cli/src/commands/analyze.rs
#	ant-cli/src/commands/file.rs
#	ant-cli/src/commands/register.rs
#	ant-cli/src/commands/vault.rs
#	ant-cli/src/commands/wallet.rs

grumbach committed:
feat: cli integration of payment re use

grumbach committed:
fix: use correct timestamp

grumbach committed:
feat: revert quote expiration to 1 hour

mickvandijke committed:
Merge pull request #2943 from grumbach/fix-dont-repay-for-failed-put

fix: don't repay for failed put

mickvandijke committed:
Merge remote-tracking branch 'maidsafe/client-light-networking' into client-light-networking-rebased

# Conflicts:
#	ant-cli/src/commands/file.rs
#	autonomi/src/client/data_types/chunk.rs
#	autonomi/src/client/payment.rs

mickvandijke committed:
fix: restore `NetworkId` usage

mickvandijke committed:
refactor(ant): remove unused optional `upload` param, `quorum`

mickvandijke committed:
fix(ant): pass `NetworkId` to `get_evm_network`

mickvandijke committed:
refactor(autonomi): get rid of dep `ant-networking`

mickvandijke committed:
fix(autonomi): libp2p missing features in import

grumbach committed:
Merge pull request #2955 from mickvandijke/client-light-networking-rebased

Client light networking rebased

grumbach committed:
Merge branch 'main' of ssh://github.com/grumbach/autonomi into client-light-networking-rebased-again

grumbach committed:
Merge pull request #2963 from grumbach/client-light-networking-rebased-again

maqi commented on autonomi/src/networking/mod.rs:
```diff
@@ -0,0 +1,326 @@
+        quorum: Quorum,
+    ) -> Result<(), NetworkError> {
+        let (tx, rx) = oneshot::channel();
+        let task = NetworkTask::PutRecord {
```
the `ant_network::put_record` carry out ChunkProofVerification as well.
However, seems with the CLN, the `verification` step got entirely skipped.

This will create a false user experience that: a success uploaded file may failed to be downloaded, due to some of chunks failed to be uploaded, say due to quote expiration.
In that situation, `kad::put_record` still return with success, but the copy won't got stored.

This step of verification can only be skipped AFTER we swtich from `kad::put_record_to` to `request/response messaging` (which will be used in the future work of `separate payment and upload`)

maqi reviewed (changes requested):
Skipping of verification could result in false user experience.
Meanwhile, the future work of `separate payment and upload` may allow skipping the verification.

Suggest to only got this PR merged AFTER bring back the verification (even on higher level), and tested with mainnet.
Otherwise, may have to wait for the `separate payment and upload` work got merged in.

maqi commented on autonomi/src/networking/driver/mod.rs:
```diff
@@ -0,0 +1,298 @@
+                    },
+                );
+            }
+            NetworkTask::GetQuote {
```
This better to be a nore generic task for `Request/Response messaging`, for the ease of other future usage.
Instead of currently too binded with `Quote`

grumbach replied:
As we don't use other `Request/Response` messaging now, this is premature in my opinion. Also, when we do add a new task using `Request/Response`, this shouldn't be a change in `NetworkTask`, but an implementation detail of how those tasks are handled. The fact that we use `Request/Response` should not bubble up and pollute the `NetworkTask` abstraction. 
Although I agree that once we do have such requests, they can use the same underlying code.

maqi commented on autonomi/src/networking/utils.rs:
```diff
@@ -0,0 +1,22 @@
+use crate::Multiaddr;
```
copyright statement

grumbach replied:
I'm not sure I understand? Could you be more specific?

maqi replied:
```
// Copyright 2025 MaidSafe.net limited.
//
// This SAFE Network Software is licensed to you under The General Public License (GPL), version 3.
// Unless required by applicable law or agreed to in writing, the SAFE Network Software distributed
// under the GPL Licence is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
// KIND, either express or implied. Please review the Licences for the specific language governing
// permissions and limitations relating to use of the SAFE Network Software.
```

maqi commented on autonomi/src/networking/driver/mod.rs:
```diff
@@ -0,0 +1,298 @@
+
+                // Add the peer addresses to our cache before sending a request.
+                for addr in &peer.addrs {
+                    self.swarm.add_peer_address(peer.peer_id, addr.clone());
```
we previously experienced many req/resp failure, and it was resolved by re-dail attempt.
will this `add_peer_address` approach achieve the same?
may need certain devnet tests or against mainnet to confirm?

grumbach replied:
I'm not sure about this one, @mickvandijke do you know about this one?

mickvandijke replied:
This worked just as good or better than dialing before sending a request when I tested.

maqi commented on autonomi/src/networking/driver/task_handler.rs:
```diff
@@ -0,0 +1,325 @@
+            NetworkTask::PutRecord { resp, .. } => {
+                self.put_record.insert(id, resp);
+            }
+            _ => {}
```
`wild card` match shall be avoided.

grumbach replied:
Oh I didn't know, sure can fix that! What was the reason for that ban again?

maqi replied:
to avoid future new entry got slipped through in some hadling process

grumbach committed:
feat: remove quote expiration

grumbach committed:
Merge pull request #2975 from grumbach/remove_quote_expiration

feat: remove quote expiration

mickvandijke committed:
fix: dial before `put_record`

mickvandijke committed:
feat(optimization): drastically increase peer cache size

mickvandijke committed:
chore: set peer cache size to 2k

mickvandijke committed:
Merge pull request #2980 from mickvandijke/fix-dial-before-put-record

fix(client): add target peers address to libp2p before `put_record_to`

maqi committed:
feat(client): upload record shall always using put_record_to

maqi committed:
feat(cli): batch entire quote-pay-upload flow

maqi committed:
Merge pull request #2978 from maqi/batched_upload

feat(cli): batch entire quote-pay-upload flow to overcome quote expiracy

maqi committed:
fix(client): avoid confusing output of private upload

maqi committed:
Merge pull request #2984 from maqi/only_use_put-record-to_for_upload

feat(client): upload record shall always using put_record_to

maqi committed:
fix(doc): correct outdated comment

maqi committed:
Merge pull request #2986 from maqi/fixing_ant-networking_API_comments

fix(doc): correct outdated comment

maqi committed:
Merge pull request #2985 from maqi/avoid_confusing_upload_printout

fix(client): avoid confusing output of private upload

grumbach committed:
fix: ignored req resp failure error

grumbach committed:
fix: compile issue on ci

mickvandijke committed:
feat: get quote / peers from 3 retries to 1

mickvandijke committed:
Merge pull request #2987 from grumbach/fix_ignored_req_resp_error

fix: ignored req resp failure error

mickvandijke committed:
Merge pull request #2988 from mickvandijke/fix-slow-client-quoting

feat: decrease get closes peers / quotes retries in client

mickvandijke committed:
Merge remote-tracking branch 'maidsafe/main' into client-light-networking

# Conflicts:
#	ant-cli/src/commands.rs
#	ant-cli/src/commands/wallet.rs

mickvandijke committed:
Merge pull request #2989 from mickvandijke/client-light-networking-rebased

client light networking rebased

maqi committed:
chore(network): remove un-used client only ant_networking methods

grumbach committed:
feat: remove more  ant networking dead code

maqi committed:
chore(network): merge get_closest_from_local methods

maqi committed:
chore(network): remove dead network_sampling code

maqi committed:
chore(network): remove unused FifoRegister

jacderida committed:
feat: support network id from 0-255

There was a bit of a misunderstanding in the way the `NetworkId` was implemented. It was reserving
the values 0, 1, and 2, for local, main, and alpha networks, respectively, then assigning 3 for any
other 'custom' network.

We change it here such that 0, 1, and 2 are reserved as above, but custom networks are permitted to
be between 3 and 255.

jacderida committed:
chore: standardise autonomi usage in connection messages

Our network is called 'Autonomi', not 'the Autonomi Network', and therefore only 'Autonomi' should
be capitalised to denote it as a proper noun.

The words 'alpha', 'local', and 'custom' are variants of Autonomi networks, so they should not be
capitalised.

Other usages of the word 'network' being capitalised were changed because 'network' is not a proper
noun in that context.

grumbach committed:
fix: eliminate put quorum errors false positives

grumbach committed:
Merge pull request #2994 from grumbach/on_quorum_error_try_get_records

mickvandijke committed:
feat: print file address within a pub archive

mickvandijke committed:
feat(ant-cli): list archive files in `files list` command

mickvandijke committed:
feat(ant-cli): better error reporting for --verbose

maqi commented on autonomi/src/networking/driver/task_handler.rs:
```diff
@@ -0,0 +1,347 @@
+                });
+
+                let res = match &records_uniq[..] {
+                    [] => responder.send(Ok((None, peers))),
```
this shall be an `error` ?
as no entry returned for a `get` ?

grumbach replied:
At the record get layer this is not an error, just no record.

maqi commented on autonomi/src/networking/driver/task_handler.rs:
```diff
@@ -0,0 +1,347 @@
+                let res = match &records_uniq[..] {
+                    [] => responder.send(Ok((None, peers))),
+                    [one] => responder.send(Ok((Some(one.clone()), peers))),
+                    [_one, _two, ..] => responder.send(Err(NetworkError::SplitRecord(holders))),
```
For mutable datatypes, situation of `split` could happen, especially with large scaled network with the mutable data got certain rounds of updates.

In such case, we shall `aggregate` the multiple versions.

Hence, here, we shall at least report back all versions

grumbach replied:
The `SplitRecord` error is managed by the datatypes layer, and is different for each type.

grumbach replied:
`NetworkError::SplitRecord(holders)` contains all versions and holders per version. The data layer can then use this data and do as it sees fit.

maqi commented on autonomi/src/networking/driver/task_handler.rs:
```diff
@@ -0,0 +1,347 @@
+                quorum,
+            }) => {
+                trace!(
+                    "QueryId({id}): GetRecordError::QuorumFailed {:?}, records: {:?}, quorum: {:?}",
```
For mutable datatypes, QuorumFailed scenario could happen with multiple versions exists.
Hence, such fetched records shall be `aggregated`.
here, at least shall report back such multiple versions.

grumbach replied:
This is up to the data layer calls to decide. The record layer has no knowledge of data types but does report splits when they happen.

maqi reviewed (changes requested):
For mutable datatypes, network get_record could fail with split or quorum_failed
but they shall be handled, instead of current `simple fail`

maqi committed:
chore(network): remove network get_record and unused config API

mickvandijke committed:
Merge pull request #2997 from mickvandijke/feat-list-individual-files-in-archive

feat: list file addr when uploading pub archive & list archive files in `files list` cmd

grumbach commented on autonomi/src/client/data_types/pointer.rs:
```diff
@@ -49,58 +53,51 @@ pub enum PointerError {
-            return Err(
-                NetworkError::RecordKindMismatch(RecordKind::DataOnly(DataTypes::Pointer)).into(),
-            );
+        {
+            Ok(Some(r)) => pointer_from_record(r)?,
+            Ok(None) => Err(GetError::RecordNotFound)?,
+            Err(NetworkError::SplitRecord(result_map)) => {
+                warn!("Pointer at {key:?} is split, trying resolution");
+                select_highest_pointer_version(result_map, key)?
+            }
+            Err(err) => {
+                error!("Error fetching pointer: {err:?}");
+                return Err(PointerError::GetError(err.into()));
+            }
         };
```
@maqi this is where is it handled for Pointers for example

maqi commented on autonomi/src/networking/driver/task_handler.rs:
```diff
@@ -0,0 +1,347 @@
+            }
+            Err(kad::GetRecordError::QuorumFailed {
+                key,
+                records,
```
this `records` returned with the failure better to be reported back ?

maqi replied:
seems the QuorumFailed error never got raised from libp2p anyway
refer further to https://github.com/libp2p/rust-libp2p/discussions/6044

jacderida committed:
chore: do not log result for `GetRecord`

Writing the result was causing excessive logging for `file download` commands.

jacderida committed:
Merge pull request #2993 from jacderida/feat-support_other_network_ids

feat: support network id from 0-255

mickvandijke committed:
feat: early return `GetRecord` if quorum is met

mickvandijke committed:
feat: gracefully terminate Kad query

jacderida committed:
Merge pull request #2990 from maqi/remove_ant-networking_client_API_methods

chore(network): remove un-used client only ant_networking methods

RolandSherwin committed:
Merge branch 'main' into client-light-networking

jacderida committed:
Merge pull request #3006 from RolandSherwin/client-light-networking

Merge main into CLN

jacderida committed:
Merge pull request #3003 from mickvandijke/optimize-download-speed

feat: early return `GetRecord` request in CLN if quorum is met

maqi committed:
fix(client): resolve deadlock issue

maqi committed:
chore(test): force wait after Update during the nodejs test

jacderida committed:
Merge pull request #3008 from maqi/fix_cli_deadlock_issue

fix(client): resolve deadlock issue
